src/main/asciidoc/index.adoc

```diff
@@ -17,503 +17,801 @@
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+In scenarios not involving a user, it's RECOMMENDED to rely on the Client Credentials grant.
+====
+
+// TODO: we should provide a comprehensive description and examples for the Client Credentials flow
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+Insecure flows such as the Resource Owner Password Credentials Grant or Implicit Grant MUST NOT be used anymore, ever.
+====
+
+// TODO: decide whether we should cover the device Authorization Flow or not
+
+== OIDC
+
+Authentication, sometimes shortened to "AuthN", is the act of confirming the identity claim of an entity.
 
-To create a valid JWS object to be able to request an access token, follow these steps:
+Built on top of OAuth 2.0 and designed with social login in mind, the OpenID Connect 1.0 (OIDC) protocol became - for
+better or for worse - the current de-facto standard for providing single sign-on (SSO) authentication, mostly as a
+consequence of the widespread adoption of OAuth 2.0.
+
+As a suite of specifications, OIDC describes an interoperable and - relatively speaking - simple identity layer built on
+top of OAuth 2.0 to perform user authentication and obtain - in a RESTful manner - basic user information in the form of
+claims contained within a JWS structure called "ID token".
+
+The ID token is intended to be used solely by clients as a medium for the retrieval of information about a user.
+
+A RESTful service is not supposed to receive or process ID tokens.
+
+As such, OIDC is outside the scope of this document.
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+Developers SHOULD nevertheless be familiar with OIDC, because understanding it usually helps in clearing up any
+ambiguities from the OAuth 2.0 specification.
+====
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+Regardless of whether OIDC is used or not, it is RECOMMENDED that a read-only /.well-known/jwks.json endpoint that
+returns the authorization server's public key set in JSON Web Key Set (JWKS) format be provided.
+====
 
-* Construct a header to define the algorithm that is used to sign the JWS.
+NOTE: A JWKS is a JSON data structure that represents a set of https://datatracker.ietf.org/doc/html/rfc7517[JSON Web
+Keys (JWK)].
+A JWK is a JSON data structure that represents a cryptographic key.
+
+This mechanism, introduced in the OpenID Connect Discovery specification, would allow resource servers to dynamically
+retrieve public keys used for token signature validation, in case key rotation is implemented.
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+A JWT MUST NOT be accepted by the resource server if it cannot identify the well-known JWKS resource in the value
+(case-sensitive string) provided by the "jku" (JWK Set URL) claim.
+====
+
+[.rule, caption="Rule {counter:rule-number}: "]
+====
+The key identifier MUST be provided in the "kid" (key identifier) claim.
```

Comment:
To avoid any ambiguity, it would be better to refer to kid as a header parameter rather than as a claim.